tools/cmaketobzl/cmaketobzl.go

```diff
@@ -0,0 +1,420 @@
+	}
+
+	switch name {
+	// TODO(shahms): Actually process these.
+	case "if", "function", "foreach", "macro":
+		counter := newCounter(name)
+		for counter.Count(name) && cmds.Advance() {
+			name = string(cmds.Head().Name)
+		}
+		return e.dispatch, nil
+	case "set":
+		e.setVariable(cmds.Head().Arguments.Eval(e.v))
+	case "unset":
+		e.unsetVariable(cmds.Head().Arguments.Eval(e.v))
+	}
+
+	if e.shouldAdd(name) {
+		args := cmds.Head().Arguments.Eval(e.v)
+		if len(args) != 1 {
+			return nil, fmt.Errorf("invalid number of arguments to directory command %s", cmds.Head().Pos)
+		}
+		if !e.excludePath(args[0]) {
+			if err := e.AddSubdirectory(cmds.Head().Arguments.Eval(e.v)[0]); err != nil {
+				return nil, err
+			}
+		}
+	}
+	cmds.Advance()
+	return e.dispatch, nil
+}
+
+// setVariable sets the value of the variable designated by the remained, following the rules of
+// https://cmake.org/cmake/help/v3.3/command/set.html#command:set
+func (e *eval) setVariable(args []string) {
+	if len(args) == 0 {
+		log.Println("Cannot set a variable without a name")
+		return
+	}
+	key, args := args[0], args[1:len(args)]
+	switch {
+	case len(args) > 0 && args[len(args)-1] == "PARENT_SCOPE":
+		e.v.SetParent(key, strings.Join(args[0:len(args)-1], ";"))
+		// When setting CACHE variables, the option can occur in either position -3 or -4 due to the optional FORCE parameter.
+	case len(args) >= 3 && args[len(args)-3] == "CACHE", len(args) >= 4 && args[len(args)-4] == "CACHE":
+		log.Println("Ignoring set of CACHE variable: ", key)
+	default:
+		e.v.Set(key, strings.Join(args, ";"))
+	}
+}
+
+// unsetVariable unsets the value of the variable designated by the remained, following the rules of
+// https://cmake.org/cmake/help/v3.3/command/set.html#command:unset
+func (e *eval) unsetVariable(args []string) {
+	switch {
+	case len(args) == 0:
+		log.Println("Cannot unset a variable without a name")
+	case len(args) == 1:
+		e.v.Set(args[0], "")
+	case len(args) == 2 && args[1] == "PARENT_SCOPE":
+		e.v.SetParent(args[0], "")
+	case len(args) == 2 && args[1] == "CACHE":
+		log.Println("Ignoring unset of CACHE variable: ", args[0])
+	default:
+		log.Println("Ignoring invalid unset command")
+	}
+}
+
+// AddSubdirectory recurses into the directory specified by dirpath and evaluates the CMakeLists.txt contained therein.
+func (e *eval) AddSubdirectory(dirpath string) error {
+	if err := e.enterDirectory(dirpath); err != nil {
+		return err
+	}
+	file, err := e.parseFile(path.Join(path.Join(e.path...), "CMakeLists.txt"))
+	if err != nil {
+		return err
+	}
+
+	cmds := commandList(file.Commands)
+	dispatch := e.dispatch
+	for {
+		dispatch, err := dispatch(&cmds)
+		if err != nil {
+			return err
+		}
+		if len(cmds) == 0 || dispatch == nil {
+			break
+		}
+
+	}
+	return e.exitDirectory(dirpath)
+}
+
+// CurrentDirectory returns the project-rooted path currently being traversed.
+func (e *eval) CurrentDirectory() string {
+	return "/" + path.Join(e.path[1:]...)
```

Comment:
Is this right? It starts with `/` suggesting it's absolute, but then omits the [0] entry, which I assume is the root?

Reply:
Yes, although it'd be nice to have more tests of this. The path returned from this is "rooted" in the project directory, but the first element of `e.path` is the common root for the paths which we want to omit.